### PR TITLE
Make connection events async

### PR DESCRIPTION
**Pull request by danielmarbach**
## Proposed Changes

Follow through on discussion in https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1675#issuecomment-2352522663

@lukebakken mentioned his [original intent was to make the events async across the board](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1675#issuecomment-2354340236) but stopped.

This PR would align the connection API surface

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [ ] I have read the `CONTRIBUTING.md` document
- [ ] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc.

### CODE REVIEW

**Comment by lukebakken** [comment]
Thank you @danielmarbach, I will review this ASAP. Just FYI, I merged #1675 and rebased your PR branch on `main`.

**Comment by danielmarbach** [comment]
Once this one is in I can change the other events too if you'd like

File: projects/RabbitMQ.Client/client/api/IConnection.cs [hunk 1]
```diff
@@ -143,9 +143,9 @@ public interface IConnection : INetworkConnection, IDisposable
         /// <see cref="IConnection"/>, then this event will be signalled whenever one
         /// of those event handlers throws an exception, as well.
         /// </remarks>
-        event EventHandler<CallbackExceptionEventArgs> CallbackException;
+        event AsyncEventHandler<CallbackExceptionEventArgs> CallbackException;
```
Comment (lukebakken) [inline projects/RabbitMQ.Client/client/api/IConnection.cs:146]: Do you think the event names should end with `Async` to be consistent with the rest of the async naming?

Comment (danielmarbach) [reply projects/RabbitMQ.Client/client/api/IConnection.cs:146]: I'm not a huge fan of the suffix, so I let you decide ;)

**Comment by lukebakken** [comment]
> Once this one is in I can change the other events too if you'd like

That would be great. From my attempt at doing this, I remember that it was the `OnSessionShutdown` event that, when made async, made everything go haywire in the test suite. That doesn't seem to be the case here (which is great).

**Comment by lukebakken** [comment]
Is now the time to add `CancellationToken` to the `AsyncEventHandler` delegate?

**Comment by danielmarbach** [comment]
> Is now the time to add `CancellationToken` to the `AsyncEventHandler` delegate?

My plan was to sweep through that after all the events are changed to be async as a dedicated PR and not mix things

File: projects/RabbitMQ.Client/client/api/IConnection.cs (hunk 1 above)
Comment (danielmarbach) [reply projects/RabbitMQ.Client/client/api/IConnection.cs]: > Yeah, let's have the suffix everywhere, to be consistent. I can get that if you'd like, just let me know.

Please go ahead

**Comment by danielmarbach** [comment]
Are you also adding the suffix to all the other events too?

**Comment by lukebakken** [comment]
> Are you also adding the suffix to all the other events too?

Everything that has changed in this PR. I'll do a commit per-rename.

File: projects/RabbitMQ.Client/client/impl/Connection.cs [hunk 2]
```diff
@@ -199,18 +209,19 @@ public event EventHandler<ConsumerTagChangedAfterRecoveryEventArgs> ConsumerTagC
...
         {
-            _callbackExceptionWrapper.Takeover(other._callbackExceptionWrapper);
-            _connectionBlockedWrapper.Takeover(other._connectionBlockedWrapper);
-            _connectionUnblockedWrapper.Takeover(other._connectionUnblockedWrapper);
-            _connectionShutdownWrapper.Takeover(other._connectionShutdownWrapper);
+            _callbackExceptionAsyncWrapper.Takeover(other._callbackExceptionAsyncWrapper);
+            _connectionBlockedAsyncWrapper.Takeover(other._connectionBlockedAsyncWrapper);
+            _connectionUnblockedAsyncWrapper.Takeover(other._connectionUnblockedAsyncWrapper);
+            _connectionShutdownAsyncWrapper.Takeover(other._connectionShutdownAsyncWrapper);
+            // TODO Why are other wrappers not taken over?
```
Comment (lukebakken) [inline projects/RabbitMQ.Client/client/impl/Connection.cs:224]: @danielmarbach my guess is that this is an oversight, I just noticed the same thing.

**Review by lukebakken — APPROVED** [review]
Renames are complete, I'll merge this once CI passes. Thanks!

**Comment by danielmarbach** [comment]
Raised the PR for the session event https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1679

**Pull request merged**